### PR TITLE
Refactor scale

### PR DESCRIPTION
**Pull request by Mozand**
#90 The axis file has been refactored and now we have axes as a function that we can import into other files.
stackedBar.ts has been refactored and utilizes the axes from another file (axis.ts)

### CODE REVIEW

**Review by MajaLinder — CHANGES_REQUESTED** [review]
Nice work Mo! I have a few comments. 

Namings changes:

- The function name "drawLine"
- Some of the axis names are from our previous namings, i.e., xScale, yScale, zScale
- I think you can remove the comments and console logs

In the picture below you can compare Tibco's bar chart with our pareto chart, and there are some things that look different:
- The ticks on the category axis should be in between the labels/bars
- The value axis and percentage scale shouldn't reach below the category axis
- The upper ticks on the value axis and percentage scale are a bit too high
- The values on the value axis are not correct
- I think we should have the percentage sign (%) after the values on the percentage scale
- Not sure about the alignment of the chart, maybe something to ask Tibco?

image

**Comment by b4rdha** [comment]
Nice progress Mo! 
I haven't had the chance to go through the code yet, but when testing I noticed some odd behaviour: 

1. it is skipping a digit after 99
2. sometimes it is showing only 00 
3. sometimes it is showing only variations of 00 and 50 

see the screenshots 

File 1
File 2
File 3

**Review by b4rdha — CHANGES_REQUESTED** [review]
Nice progress Mo! 
I haven't had the chance to go through the code yet, but when testing I noticed some odd behaviour: 

1. it is skipping a digit after 99
2. sometimes it is showing only 00 
3. sometimes it is showing only variations of 00 and 50 

see the screenshots 

File 1
File 2
File 3

File: src/axis.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import * as d3 from "d3";
+import { Pareto} from "./pareto";
+
+ export async function drawLine(pareto: Pareto) {
```
Comment (b4rdha) [inline src/axis.ts:4]: I think naming here is misleading, since this draws the axes, drawAxes is more appropriate

File: src/axis.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    d3.selectAll("path").remove();
+    d3.selectAll("g").remove();
+
+    const maxValueHeight:number = pareto.maxValue;
```
Comment (b4rdha) [inline src/axis.ts:8]: paretoMaxValue perhaps, since it's not really height

File: src/axis.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    const maxValueWidth:any = [];
```
Comment (b4rdha) [inline src/axis.ts:9]: since it's an array, it should be called paretoCategoryValues

File: src/axis.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    pareto.stackedBars.forEach((p) => {
+        maxValueWidth.push(p.label);
+    });
+  
+    var svg = d3.select("svg");
+    const valueAxisHeight:any = document.querySelector("#svg");
```
Comment (b4rdha) [inline src/axis.ts:15]: this also, its  a container, not returning any value, so naming is misleading

File: src/axis.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    const valuePixels:any = valueAxisHeight.getBoundingClientRect();
```
Comment (b4rdha) [inline src/axis.ts:16]: I would name it as svgBoundingClientRect

File: src/axis.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    const categoryAxisWidth:any = document.querySelector("#svg");
```
Comment (b4rdha) [inline src/axis.ts:17]: you already called this above, so it's redundant

File: src/axis.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    const categoryPixels:any = categoryAxisWidth.getBoundingClientRect();
```
Comment (b4rdha) [inline src/axis.ts:18]: ditto

File: src/axis.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    console.log(categoryPixels.width);
```
Comment (b4rdha) [inline src/axis.ts:19]: so console will become svgBoundingClientRect.width

File: src/axis.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    console.log(valuePixels.height);
```
Comment (b4rdha) [inline src/axis.ts:20]: also as above

File: src/axis.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    var svg = d3.select("svg");
+// Step 4
+    var xScale = d3
+            .scaleBand()
```
Comment (b4rdha) [inline src/axis.ts:24]: and then the ramifications below also change as a consequence

**Review by MajaLinder — CHANGES_REQUESTED** [review]

**Comment by MajaLinder** [comment]
Actually @b4rdha, the odd skipping behaviour you found seems to be related to the positioning of the scale. If it's moved a little to the right, the numbers appear. So I believe the issue is that the numbers are cut-off because the scale is positioned too much to the left.

File: src/axis.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        yScale = d3
+            .scaleLinear()
+            .domain([0, maxValueHeight])
+            .range([valuePixels.height - 50, 3]),
+        zScale = d3
+            .scaleLinear()
+            .domain([0, 100])
+            .range([valuePixels.height - 50, 3]);
+
+    var g = svg
```
Comment (MajaLinder) [inline src/axis.ts:36]: Why do you rename the svg here?

File: src/axis.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        .append("g") //move bars
```
Comment (MajaLinder) [inline src/axis.ts:37]: What is meant by "move bars"?

File: src/axis.ts (hunk 5 above)
Comment (Mozand) [reply src/axis.ts:16]: it returns pixels I think this is self-explanatory

**Review by b4rdha — APPROVED** [review]
I think it's working now, good job! 

Comparing it with Tibco's visualisation: 
I think that there's too much unused space in the right. 
And there might be an issue with the scale values, I think they are too frequent. 

But that can be refined I suppose on the next sprint. 
Screenshot 2022-11-17 at 13 06 46



I didn't re-check the previously requested changes and their status is not reflected here in the PR either.  I will assume they have been fixed and approve the PR.

**Review by MajaLinder — APPROVED** [review]

**Comment by b4rdha** [comment]
Also leaving this picture of the spaces, fur further consideration
Screenshot 2022-11-17 at 13 45 52

**Pull request merged**